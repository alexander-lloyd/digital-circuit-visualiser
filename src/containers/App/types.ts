import {
    SET_SOURCE_REQUEST,
    SET_SOURCE_SUCCESS,
    SET_SOURCE_FAILURE,
    RESET_SCALE,
    ZOOM_IN,
    ZOOM_OUT
} from './constants';
import {AST} from 'lib/parser/ast';

export type DispatchFunction = (action: AppActions) => void;

/**
 * Set the source code of the canvas.
 */
export interface SetSourceCodeRequestAction {
    type: typeof SET_SOURCE_REQUEST;
    source: string;
}

/**
 * Code successfully compiled. Set the AST.
 */
export interface SetSourceCodeSuccessAction {
    type: typeof SET_SOURCE_SUCCESS;
    ast: AST;
}

/**
 * Set the source code of the canvas.
 */
export interface SetSourceCodeFailureAction {
    type: typeof SET_SOURCE_FAILURE;
    reason: string;
}

/**
 * Reset Scale Action.
 */
export interface ResetScaleAction {
    type: typeof RESET_SCALE;
}

/**
 * Zoom In Action.
 */
export interface ZoomInAction {
    type: typeof ZOOM_IN;
}

/**
 * Zoom Out Action.
 */
export interface ZoomOutAction {
    type: typeof ZOOM_OUT;
}

/**
 * Canvas Actions.
 */
export type AppActions =
    | SetSourceCodeRequestAction
    | SetSourceCodeSuccessAction
    | SetSourceCodeFailureAction
    | ResetScaleAction
    | ZoomInAction
    | ZoomOutAction;

/**
 * Canvas Container State.
 */
export interface GlobalState {
    download: {
        loading: boolean;
    };
    scale: number;
    code: string;
    // On initial load we won't have an AST until success Action.
    ast: AST | null;
    // No error string on successful Action.
    errorString: string | null;
}
