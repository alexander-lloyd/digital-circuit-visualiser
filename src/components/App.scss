.fullscreen {
    height: 100%;
    width: 100%;
}

.fullheight {
    height: 100%;
}