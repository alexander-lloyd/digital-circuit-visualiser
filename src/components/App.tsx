import * as React from 'react';

export const App = () => <p>Hello World</p>;

