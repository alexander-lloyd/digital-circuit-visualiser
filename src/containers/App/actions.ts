import {
    MODAL_HIDE,
    MODAL_SHOW,
    SET_FEATURE_FLAG,
    SET_SOURCE_REQUEST,
    SET_SOURCE_SUCCESS,
    SET_SOURCE_FAILURE,
    RESET_SCALE,
    ZOOM_IN,
    ZOOM_OUT
} from './constants';
import {
    DispatchFunction,
    ModalHideAction,
    ModalShowAction,
    SetFeatureFlagAction,
    SetSourceCodeRequestAction,
    SetSourceCodeSuccessAction,
    SetSourceCodeFailureAction,
    ResetScaleAction,
    ZoomInAction,
    ZoomOutAction
} from './types';
import {compile, SyntaxError, AST} from '../../lib/parser/index';
import {FEATURES_KEYS} from '../../assets/features';

/**
 * Create a ModalHideAction.
 *
 * @returns ModalHideAction.
 */
export function modalHideAction(): ModalHideAction {
    return {
        type: MODAL_HIDE
    };
}

/**
 * Create a ModalShowAction.
 *
 * @returns ModalShowAction.
 */
export function modalShowAction(): ModalShowAction {
    return {
        type: MODAL_SHOW
    };
}

/**
 * Set a feaure flag.
 *
 * @param name Feature Name.
 * @param value Set feature value.
 * @returns SetFeatureFlagAction.
 */
export function setFeatureFlag(name: FEATURES_KEYS, value: boolean): SetFeatureFlagAction {
    return {
        type: SET_FEATURE_FLAG,
        name,
        value
    };
}

/**
 * Set the source code. Fires off Actions.
 *
 * @param dispatch Dispatch Function.
 * @returns Action Creater with side effect.
 */
export function setSourceCode(dispatch: DispatchFunction): (source: string) => void {
    return (source: string): void => {
        dispatch({
            type: SET_SOURCE_REQUEST,
            source
        });
        // Compile.
        try {
            const ast = compile(source);
            dispatch({
                type: SET_SOURCE_SUCCESS,
                ast
            });
        } catch (e) {
            const reason = (e as SyntaxError).message;
            dispatch({
                type: SET_SOURCE_FAILURE,
                reason
            });
        }
    };
}

/**
 * Request to set the Source Code.
 *
 * @param source Source Code.
 * @returns Set Source Code Request Action.
 */
export function setSourceCodeRequest(source: string): SetSourceCodeRequestAction {
    return {
        type: SET_SOURCE_REQUEST,
        source
    };
}

/**
 * Request to set the AST in the state.
 *
 * @param ast AST.
 * @returns Set Source Code Success Action.
 */
export function setSourceCodeSuccess(ast: AST): SetSourceCodeSuccessAction {
    return {
        type: SET_SOURCE_SUCCESS,
        ast
    };
}

/**
 * Fired if parsing failed.
 *
 * @param reason Failure Reason.
 * @returns Set Source Code Request Action.
 */
export function setSourceCodeFailure(reason: string): SetSourceCodeFailureAction {
    return {
        type: SET_SOURCE_FAILURE,
        reason
    };
}

/**
 * Reset Zoom Action Creater.
 *
 * @returns Reset Scale Action.
 */
export function resetZoom(): ResetScaleAction {
    return {
        type: RESET_SCALE
    };
}

/**
 * Zoom In Action Creater.
 *
 * @returns Zoom In Action.
 */
export function zoomIn(): ZoomInAction {
    return {
        type: ZOOM_IN
    };
}

/**
 * Zoom Out Action Creater.
 *
 * @returns Zoom Out Action.
 */
export function zoomOut(): ZoomOutAction {
    return {
        type: ZOOM_OUT
    };
}

/**
 * Canvas Action Creaters.
 */
export interface ActionCreaters {
    setSourceCode: typeof setSourceCodeRequest;
    resetZoom: typeof resetZoom;
    zoomIn: typeof zoomIn;
    zoomOut: typeof zoomOut;
}
