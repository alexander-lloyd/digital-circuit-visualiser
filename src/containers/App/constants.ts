// Actions
export const SET_SOURCE_REQUEST = 'SET_SOURCE_REQUEST';
export const SET_SOURCE_SUCCESS = 'SET_SOURCE_SUCCESS';
export const SET_SOURCE_FAILURE = 'SET_SOURCE_FAILURE';
export const ZOOM_IN = 'ZOOM_IN';
export const ZOOM_OUT = 'ZOOM_OUT';
export const RESET_SCALE = 'RESET_SCALE';
export const DOWNLOAD_SUCCESS = 'DOWNLOAD_SUCCESS';
