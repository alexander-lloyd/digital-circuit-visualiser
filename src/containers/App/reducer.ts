import {
    RESET_SCALE,
    SET_SOURCE_REQUEST,
    SET_SOURCE_SUCCESS,
    SET_SOURCE_FAILURE,
    ZOOM_IN,
    ZOOM_OUT
} from './constants';
import {AppActions, GlobalState} from './types';
import {
    INITIAL_CODE,
    INITIAL_SCALE,
    MAXIMUM_SCALE,
    MINIMUM_SCALE,
    SCALING_FACTOR
} from '../Canvas/constants';


const initialState: GlobalState = {
    download: {
        loading: false
    },
    scale: INITIAL_SCALE,
    code: INITIAL_CODE,
    ast: null,
    errorString: null
};

/**
 * Canvas Reducer.
 *
 * @param state Previous State.
 * @param action Canvas Action.
 * @returns New State.
 */
export function reducer(state = initialState, action: AppActions): GlobalState {
    const {scale: oldScale} = state;
    console.log(`[GLOBAL] ${action.type}`, action);

    switch (action.type) {
    case SET_SOURCE_REQUEST: {
        const {source} = action;
        if (source === state.code) {
            return state;
        }
        return {
            ...state,
            code: source
        };
    }
    case SET_SOURCE_SUCCESS: {
        const {ast} = action;
        return {
            ...state,
            ast,
            // We don't have an error.
            errorString: null
        };
    }
    case SET_SOURCE_FAILURE: {
        const {reason} = action;
        return {
            ...state,
            // Set the AST to null?
            ast: null,
            errorString: reason
        };
    }
    case RESET_SCALE: {
        return {
            ...state,
            scale: INITIAL_SCALE
        };
    }
    case ZOOM_IN: {
        const newScale = Math.min(oldScale * SCALING_FACTOR, MAXIMUM_SCALE);
        return {
            ...state,
            scale: newScale
        };
    }
    case ZOOM_OUT: {
        const newScale = Math.max(oldScale / SCALING_FACTOR, MINIMUM_SCALE);
        return {
            ...state,
            scale: newScale
        };
    }
    default:
        return state;
    }
}
