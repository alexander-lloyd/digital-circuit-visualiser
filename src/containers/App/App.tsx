import React, {useEffect} from 'react';
import {connect} from 'react-redux';

import Canvas from '../Canvas/index';
import Header from '../../components/Header';
import Modal from '../../components/Modal';
import Sidebar from '../../components/Sidebar';

import * as actions from './actions';
import {INITIAL_CODE} from './constants';
import {GlobalState, DispatchFunction} from './types';

import {HELP_TITLE, HelpContentComponent} from '../../assets/help';

import './App.scss';

/**
 * App Component State.
 */
interface AppState {
    showModal: boolean;
}

/**
 * App Actions.
 */
interface AppDispatchProps {
    closeModal: () => void;
    setInitialSourceCode: () => void;
}

/**
 * App Properties.
 */
interface AppProps extends AppState, AppDispatchProps {}

/**
 * App Component.
 *
 * @returns App
 */
export function App({showModal, closeModal, setInitialSourceCode}: AppProps): JSX.Element {
    useEffect(() => {
        setInitialSourceCode();
    });

    return (
        <div className="fullscreen">
            <Header />
            <div className="columns is-fullheight-with-navbar is-gapless"
                 style={{height: 'calc(100% - 1rem)'}}>
                <div className="column fullheight is-narrow">
                    <Sidebar />
                </div>
                <div className="column">
                    <Canvas />
                </div>
            </div>
            <Modal closeAction={closeModal}
                   show={showModal}
                   title={HELP_TITLE} >
                <HelpContentComponent />
            </Modal>
        </div>
    );
}


/**
 * Map state to props.
 *
 * @returns Component Props.
 */
function mapStateToProps({showModal}: GlobalState): AppState {
    return {
        showModal
    };
}

/**
 * Map Dispatch to Props.
 *
 * @param dispatch Action dispatcher.
 * @returns Props.
 */
function mapDispatchToProps(dispatch: DispatchFunction): AppDispatchProps {
    const {modalHideAction, setSourceCode} = actions;
    return {
        closeModal: (): void => dispatch(modalHideAction()),
        setInitialSourceCode: (): void => setSourceCode(dispatch)('A . B . C')
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(App);
