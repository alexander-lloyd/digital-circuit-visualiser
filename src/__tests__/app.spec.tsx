import React from 'react';
import {shallow} from 'enzyme';

import {App} from '../containers/App/App';

describe('app Component', () => {
    it('should render without errors', () => {
        expect.assertions(1);
        /* eslint-disable-next-line jsdoc/require-jsdoc, @typescript-eslint/no-empty-function */
        const closeModalAction = (): void => {};
        const wrapper = shallow(<App closeModal={closeModalAction}
                                     showModal={false} />);

        expect(wrapper).toMatchInlineSnapshot('ShallowWrapper {}');
    });
});
