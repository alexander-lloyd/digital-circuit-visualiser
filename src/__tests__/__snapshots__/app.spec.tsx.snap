// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render without errors 1`] = `ShallowWrapper {}`;
